### PR TITLE
Create an inherited conformance in SubstitutionMap::lookupConformance when needed.

### PR DESCRIPTION
**Pull request by jckarter**
If the substituted type for a conformance found via a superclass constraint is a subclass of that
superclass, then we should represent that with an InheritedProtocolConformance rather than with
the original root conformance that applies to the superclass. If we don't do this, then we end up
with spurious inequalities in generic signatures that ought to be equivalent, because some
paths use the inherited conformance and some don't, as in SR-12330 | rdar://problem/60174186.

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

File: lib/AST/SubstitutionMap.cpp [hunk 1]
```diff
@@ -355,8 +355,15 @@ SubstitutionMap::lookupConformance(CanType type, ProtocolDecl *proto) const {
   // Check whether the superclass conforms.
   if (auto superclass = genericSig->getSuperclassBound(type)) {
     LookUpConformanceInSignature lookup(getGenericSignature().getPointer());
-    if (auto conformance = lookup(type->getCanonicalType(), superclass, proto))
+    if (auto conformance = lookup(type->getCanonicalType(), superclass, proto)){
```
Comment (slavapestov) [inline lib/AST/SubstitutionMap.cpp]: LookUpConformanceInSignature just does a ModuleDecl::lookupConformance(). Why doesn't that do the right thing? Should we instead just `s/type/substType/` here?

Comment (jckarter) [reply lib/AST/SubstitutionMap.cpp]: It seems to do more than that. `type` here appears to intentionally be referring to the unsubstituted type.

Comment (slavapestov) [reply lib/AST/SubstitutionMap.cpp]: Ah, right.

So I think instead you want `s/superclass/substType/` in the `lookup()` call.

`LookUpConformanceInSignature::operator()` ignores `dependentType` (the original type) and calls `GenericSignature::lookupConformance()` with `conformingReplacementType` (the substituted type).

The issue is that we're calling `operator()` (via `lookup()`) with `superclass` as the `conformingReplacementType`. I think `superclass` should be `substType` here.

Comment (jckarter) [reply lib/AST/SubstitutionMap.cpp]: OK, I'll try that.

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

**Comment by jckarter** [comment]
@slavapestov Does this look better?

File: lib/AST/Type.cpp [hunk 2]
```diff
@@ -2447,6 +2447,16 @@ getForeignRepresentable(Type type, ForeignLanguage language,
         break;
       }
     }
+    
+    // Specialize the conformance we were given for the type we're testing.
+    if (result.getKind() == ForeignRepresentableKind::Bridged
```
Comment (slavapestov) [inline lib/AST/Type.cpp]: This change looks like it's related to Objective-C bridging, is there a test case for it?

Comment (jckarter) [reply lib/AST/Type.cpp]: I caught it incidentally while trying to build with my `-verify-all-substitution-maps` flag I added. I didn't reduce a test case.

Comment (slavapestov) [reply lib/AST/Type.cpp]: Alright, no worries.

**Pull request merged**